### PR TITLE
Miroil

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Miroil: Lomiri compatibility library

The plan:

1. Copies those elements from QtMir that historically rely on libmirserver
2. Copies those elements from unity-system-compositor that historically rely on libmirserver
3. Provides Wayland extension(s) that replicate functionality that was lost when mirclient was deprecated

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Latest iteration featured here: https://forums.ubports.com/topic/7237/miroil-for-20-04/

File: include/miroil/miroil/mirbuffer.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIROIL_GLBUFFER_H
+#define MIROIL_GLBUFFER_H
+
+#include <mir/geometry/size.h>
+#include <mir/version.h>
```
Comment (wmww) [inline include/miroil/miroil/mirbuffer.h:21]: Is this still required?

File: include/miroil/miroil/mirbuffer.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    explicit GLBuffer(std::shared_ptr<mir::graphics::Buffer> const& buffer);
+
+    operator bool() const;
+    bool has_alpha_channel() const;
+    mir::geometry::Size size() const;
+
+    void reset();
+    void reset(std::shared_ptr<mir::graphics::Buffer> const& buffer);
+    void bind();
+#if MIR_SERVER_VERSION < MIR_VERSION_NUMBER(2, 3, 0)
```
Comment (wmww) [inline include/miroil/miroil/mirbuffer.h:44]: Is this check still required?

File: include/miroil/miroil/mirbuffer.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    void gl_bind_tex();
+#endif
+
+private:
+    void init();
+    void destroy();
+
+    std::shared_ptr<mir::graphics::Buffer> wrapped;
+    GLuint m_textureId;
+#if MIR_SERVER_VERSION < MIR_VERSION_NUMBER(2, 3, 0)
```
Comment (wmww) [inline include/miroil/miroil/mirbuffer.h:54]: Or this?

File: src/miroil/input_device.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+}
+        
+void miroil::InputDevice::apply_keymap(std::string const& layout, std::string const& variant)
+{
+    using namespace mir::input;
+
+    if (auto const kbd_conf = device->keyboard_configuration())
+    {
+        // TODO work out how to support this functionality on OSK etc.
+#if MIR_SERVER_VERSION >= MIR_VERSION_NUMBER(2, 5, 0)
```
Comment (wmww) [inline src/miroil/input_device.cpp:48]: Is this check required?

File: src/miroil/mir_connection.cpp [hunk 5]
```diff
@@ -0,0 +1 @@
+
```
Comment (wmww) [inline src/miroil/mir_connection.cpp:1]: Empty file

File: src/miroil/mir_server_hooks.cpp [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    
+private:    
+    std::shared_ptr<miroil::InputDeviceObserver> observer;
+};
+
+struct HiddenCursorWrapper : mg::Cursor
+{
+    HiddenCursorWrapper(std::shared_ptr<mg::Cursor> const& wrapped) :
+        wrapped{wrapped} { wrapped->hide(); }
+#if MIR_SERVER_VERSION < MIR_VERSION_NUMBER(2, 3, 0)
```
Comment (wmww) [inline src/miroil/mir_server_hooks.cpp:72]: Another version check

File: src/miroil/mirbuffer.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+miroil::GLBuffer::GLBuffer() = default;
+miroil::GLBuffer::~GLBuffer()
+{
+    destroy();
+}
+
+miroil::GLBuffer::GLBuffer(std::shared_ptr<mir::graphics::Buffer> const& buffer) :
+    wrapped(buffer),
+    m_textureId(0)
+#if MIR_SERVER_VERSION < MIR_VERSION_NUMBER(2, 3, 0)
```
Comment (wmww) [inline src/miroil/mirbuffer.cpp:38]: This file has more version checks

File: src/miroil/wayland-protocols/mir-trusted-prompt-sessions-v1.xml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+  </copyright>
+
+  <description summary="Trusted prompt session protocol">
+    TODO
```
Comment (wmww) [inline src/miroil/wayland-protocols/mir-trusted-prompt-sessions-v1.xml:27]: This should probably be filled in

**Review by wmww — CHANGES_REQUESTED** [review]
Are all the `#ifdef` version checks still required? I figured they wouldn't be once this is merged into Mir, so I commented on all the ones I saw.

File: src/miroil/input_device.cpp (hunk 4 above)
Comment (AlanGriffiths) [reply src/miroil/input_device.cpp:48]: It makes for easier syncing with a mir-1.8 codebase

File: src/miroil/wayland-protocols/mir-trusted-prompt-sessions-v1.xml (hunk 8 above)
Comment (AlanGriffiths) [reply src/miroil/wayland-protocols/mir-trusted-prompt-sessions-v1.xml:27]: That was (and is) the plan. But it hasn't happened yet.

File: src/miroil/mir_connection.cpp (hunk 5 above)
Comment (AlanGriffiths) [reply src/miroil/mir_connection.cpp:1]: Gone!

**Comment by AlanGriffiths** [comment]
> Are all the `#ifdef` version checks still required? I figured they wouldn't be once this is merged into Mir, so I commented on all the ones I saw.

They are (or something equivalent). The plan downstream is to sync this code into QtMir for building against Mir 1.x and these avoid a few spurious differences.

**Comment by wmww** [comment]
Ah, so there is a use for the version checks. In that case they're fine.

Event builder is acceptable to me.

Only thing left is to write prompt session protocol documentation.

**Comment by AlanGriffiths** [comment]
> Only thing left is to write prompt session protocol documentation.

I'm intentionally leaving that as a "TODO".

**Pull request merged**